Paginate edxnotes frontend

muhammad-ammar opened the pull request:
## [TNL-3908](https://openedx.atlassian.net/browse/TNL-3908)

This adds the pagination to  the edx-notes frontend.

Sandbox: https://notes-pagination.sandbox.edx.org

### Testing
- [X] Jasmine
- [x] Bokchoy

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @symbolist 
- [x] Code review: @muzaffaryousaf
- [x] UX review: @talbs @chris-mike
- [x] a11y review: @cptvitamin 
- [x] Product review: @explorerleslie 

### Post-review
- [x] Squash commits

muzaffaryousaf commented on lms/static/js/edxnotes/views/page_factory.js:
```diff
@@ -6,19 +6,28 @@ define([
+            {
+                url: params.notesEndpoint,
+                perPage: params.pageSize,
+                parse: true
```
Why we need to pass `parse:true` ??

muhammad-ammar replied:
Collection will not parse the results and create models if we will not set `parse:true`

muzaffaryousaf commented:
@muhammad-ammar Can we handle error case for this page, for now if i get an error from server side, it only appears in browser console.

muzaffaryousaf commented:
@talbs Can you please review this PR from UX perspective. We have added the pagination to the student notes, you can see them in [Notes tab](https://notes-pagination.sandbox.edx.org/courses/course-v1:edX+Test101+course/edxnotes/). Also, i've added enough notes to see the pagination for staff user. Thanks.

@cptvitamin We have added pagination to [student notes](https://notes-pagination.sandbox.edx.org/courses/course-v1:edX+Test101+course/edxnotes/) please give it a look in a11y perspective.

@talbs & @cptvitamin FYI, we only added the pagination to notes in "Recent Activity" sub tab & for search results. Please let me know if you need anything else. Thanks.

muzaffaryousaf commented on lms/static/js/spec/edxnotes/collections/notes_spec.js:
```diff
@@ -23,11 +23,22 @@ define([
+            var compareUnits = function (structureUnits, collectionUnits) {
+                expect(structureUnits.length === collectionUnits.length).toBeTruthy();
+                for(var i = 0; i < structureUnits.length; i++) {
+                    expect(structureUnits[i].attributes).toEqual(collectionUnits[i].attributes)
```
semi-colon missing.

muzaffaryousaf commented on lms/static/js/spec/edxnotes/collections/notes_spec.js:
```diff
@@ -23,11 +23,22 @@ define([
+            };
+
+            _.each(units, function(value, key){
+                compareUnits(structure.units[key], value)
```
semi-colon missing.

muzaffaryousaf commented on lms/static/js/spec/edxnotes/helpers.js:
```diff
@@ -106,57 +108,135 @@ define(['underscore'], function(_) {
+    };
+
+    stringStartsWith = function (string, prefix) {
+        return string.slice(0, prefix.length) == prefix;
```
quality is failing for this. Please use `===` instead.

muzaffaryousaf commented on lms/static/js/spec/edxnotes/views/search_box_spec.js:
```diff
@@ -1,14 +1,25 @@
     'use strict';
     describe('EdxNotes SearchBoxView', function() {
-        var getSearchBox, submitForm, assertBoxIsEnabled, assertBoxIsDisabled;
+        var getSearchBox, submitForm, assertBoxIsEnabled, assertBoxIsDisabled, searchReponse;
```
`searchResponse`

muzaffaryousaf commented on lms/static/js/spec/edxnotes/views/search_box_spec.js:
```diff
@@ -1,14 +1,25 @@
-        var getSearchBox, submitForm, assertBoxIsEnabled, assertBoxIsDisabled;
+        var getSearchBox, submitForm, assertBoxIsEnabled, assertBoxIsDisabled, searchReponse;
+
+        searchReponse = {
```
Same as above.

muzaffaryousaf commented on lms/static/js/spec/edxnotes/helpers.js:
```diff
@@ -169,6 +249,13 @@ define(['underscore'], function(_) {
         getUnit: getUnit,
-        getDefaultNotes: getDefaultNotes
+        getDefaultNotes: getDefaultNotes,
+        stringStartsWith: stringStartsWith,
```
I think we are not using this method anywhere else. If that's the case then we can remove this from here.

muzaffaryousaf commented on lms/static/js/spec/edxnotes/helpers.js:
```diff
@@ -106,57 +108,135 @@ define(['underscore'], function(_) {
+    };
+
+    verifyUrl = function (requestUrl, expectedUrl, expectedParams) {
+        expect(stringStartsWith(requestUrl, expectedUrl)).toBeTruthy();
```
This method is being used here only; If that's the case, we can replace this call with actual 1 line statement.

muzaffaryousaf commented on lms/static/js/spec/edxnotes/helpers.js:
```diff
@@ -106,57 +108,135 @@ define(['underscore'], function(_) {
+        return string.slice(0, prefix.length) == prefix;
+    };
+
+    verifyUrl = function (requestUrl, expectedUrl, expectedParams) {
```
This method is not only verifying the url but also request params, can we name this to `verifyUrlAndRequestParams` Or is it possible to merge both methods `verifyUrl` and `verifyRequestParams`?

muhammad-ammar replied:
We can merge them to a single function but i will prefer to remain them as is because we are only verifying request params in some tests.

muzaffaryousaf commented on lms/static/js/spec/edxnotes/helpers.js:
```diff
@@ -106,57 +108,135 @@ define(['underscore'], function(_) {
+            current_page: options.current_page || 1,
+            start: options.start || 0,
+            results: []
+        };
```
Don't we need next and previous ?

muhammad-ammar replied:
Nope. next and previous urls are not used by paginator. When next or previous button is clicked, paginator calculates the new page number from current page and then sends a get request to fetch the results for that page.

muzaffaryousaf commented on lms/static/js/spec/edxnotes/views/tabs/recent_activity_spec.js:
```diff
@@ -50,28 +70,135 @@ define([
+        it("will not render header and footer if there are no notes", function () {
+            var notes = Helpers.createNotesData(
+                {
+                    numNotesToCreate: 0,
```
Why we are calling this method with `numNotesToCreate = 0` ? I think we don't need this call.

muzaffaryousaf replied:
Just to get data ? That does not make sense to call create note with 0 length.

muzaffaryousaf replied:
Or is it due to the default notes ?

muzaffaryousaf commented on lms/static/js/spec/edxnotes/views/tabs/recent_activity_spec.js:
```diff
@@ -50,28 +70,135 @@ define([
+            var collection = new NotesCollection(notes, {perPage: 10, parse: true});
+            var view = getView(collection, this.tabsCollection);
+
+            Helpers.verifyPaginationInfo(view, "Showing 1-10 out of 12 total", 1, 2);
```
If we are passing 10 notes to create, how we ended up with this number ?

muzaffaryousaf replied:
I think that's due to count, is that true ? If that's the case can we make this consistent with num of notes ?

muzaffaryousaf replied:
Or is it due to the default notes ?

muhammad-ammar replied:
`numNotesToCreate` is the number of notes in the current page in response and `count` will be the total number of notes available. We can't make them same.

muzaffaryousaf commented on lms/static/js/spec/edxnotes/views/tabs/search_results_spec.js:
```diff
@@ -195,20 +188,140 @@ define([
+        it("will not render header and footer if there are no notes", function () {
+            var view = getView(this.tabsCollection),
+                requests = AjaxHelpers.requests(this),
+                notes = Helpers.createNotesData(
```
If no notes then we don't need to call this.

muzaffaryousaf commented:
@muhammad-ammar I'm done with first pass. Let me know when you are done with changes. Thanks.

muhammad-ammar commented:
@muzaffaryousaf I have addressed your feedback. Please continue the review. Thanks

muhammad-ammar commented:
> Can we handle error case for this page, for now if i get an error from server side, it only appears in browser console.

I am not sure what is the requirement here but the current pattern is not handle the error occurred during page navigation. This is the same approach happening in Teams and Bookmarks.

muhammad-ammar commented:
@muzaffaryousaf ^^

symbolist commented on lms/djangoapps/edxnotes/helpers.py:
```diff
@@ -322,13 +322,21 @@ def get_notes(request, course, page=DEFAULT_PAGE, page_size=DEFAULT_PAGE_SIZE, t
         raise EdxNotesParseError(_("Invalid response received from notes api."))
```
Can we log this error?

symbolist replied:
Also the other exceptions.

symbolist commented on lms/static/js/edxnotes/collections/notes.js:
```diff
@@ -1,11 +1,29 @@
+
+            // These are not specified for the Notes API
+            delete this.server_api.sort_order;
+            delete this.server_api.text_search;
```
A shorter and more robust version will be:
```
this.server_api = _.pick(PagingCollection.prototype.server_api, "page", "page_size");
if (options.text) { this.server_api['text'] = options.text };
```

symbolist commented on lms/static/js/edxnotes/views/notes_page.js:
```diff
@@ -31,7 +31,8 @@ define([
                 tabsCollection: this.tabsCollection,
-                scrollToTag: scrollToTag
+                scrollToTag: scrollToTag,
+                createHeaderFooter: true
```
Do we also need to pass `perPage`?

muhammad-ammar replied:
nope. we only need to pass the `perPage` to `searchResultsView` and from there it will be passed to `SearchBoxView` to initialize the notes collection.

symbolist commented on lms/djangoapps/edxnotes/helpers.py:
```diff
@@ -322,13 +322,21 @@ def get_notes(request, course, page=DEFAULT_PAGE, page_size=DEFAULT_PAGE_SIZE, t
+    # 2. directly call the notes api
+    # In first case, notes frontend requires count and results attributes in json response and
+    # for second case frontend requires total and rows attributes in json response.
+    collection['count'] = collection['total']
```
```
# Notes API is called from:
# 1. The annotator in courseware. It expects these attributes to be named "total" and "rows".
# 2. The Notes tab Javascript proxied through LMS. It expects these attributes to be called "count" and "results".
```

symbolist commented on lms/static/js/edxnotes/views/page_factory.js:
```diff
@@ -6,19 +6,28 @@ define([
             collection: collection,
             debug: params.debugMode,
-            endpoint: params.endpoint
+            endpoint: params.endpoint,
```
I only see `notesEndpoint` being passed to the factory from the template. Typo?

symbolist replied:
I still do not understand the difference between `endpoint` and `notesEndpoint`.

muhammad-ammar replied:
@symbolist I don't see `endpoint` passed from edxnotes django view so i have removed it.

symbolist commented on lms/static/js/edxnotes/views/tab_panel.js:
```diff
@@ -13,14 +14,30 @@ function (gettext, _, Backbone, NoteItemView) {
         },
 
+        renderView: function(view) {
+            if (this.options.createHeaderFooter && this.collection.models.length) {
```
Why the `this.collection.models.length`?

muhammad-ammar replied:
If there is no data then don't show the header and footer.

symbolist commented on lms/static/js/spec/edxnotes/views/tabs/search_results_spec.js:
```diff
@@ -195,20 +188,140 @@ define([
+            Helpers.respondToRequest(requests, page1Notes);
+
+            Helpers.verifyPaginationInfo(view, "Showing 1-10 out of 15 total", 1, 2);
+            Helpers.verifyPageData(view, this.tabsCollection, tabInfo, searchResultsTabId, page1Notes);
```
A lot of the logic in these tests is the same as the logic in the new tests in `recent_activity_spec.js` Can we not make helper functions and just pass in the different parameters for the two views?

symbolist commented:
@muhammad-ammar I am done with the first pass.

muhammad-ammar commented:
@talbs @cptvitamin Please hold on your reviews for now. We may need to make some UI changes.

muhammad-ammar commented:
@explorerleslie Please take a look. We have removed the `Location in Course` and `Tags` tabs. Thanks!

explorerleslie commented:
:+1:

muhammad-ammar commented:
@talbs @cptvitamin Please continue the review. Thanks!

muhammad-ammar commented:
@talbs @clytwynec We have added pagination to [edxnotes](https://notes-pagination.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/edxnotes/) and removed `Location in Course` and `Tags` tabs. Although edxnotes and paginator controls were already reviewed for accessibility but we want to make sure that we haven't missed anything.

muzaffaryousaf commented:
jenkins run lettuce

muzaffaryousaf commented:
jenkins run bokchoy

muzaffaryousaf commented:
jenkins run lettuce

muzaffaryousaf commented:
jenkins run lettuce

muzaffaryousaf commented on common/test/acceptance/tests/lms/test_lms_edxnotes.py:
```diff
@@ -884,7 +932,7 @@ def test_scroll_to_tag_search(self):
         self._scroll_to_tag_and_verify("pumpkin", 4)
 
-    def test_scroll_to_tag_from_tag_view(self):
+    def xtest_scroll_to_tag_from_tag_view(self):
```
@muhammad-ammar Can you explain here why we need to skip these ?

muzaffaryousaf replied:
Also, can we use @skip with reason ?

muzaffaryousaf commented on lms/djangoapps/edxnotes/helpers.py:
```diff
@@ -106,6 +106,7 @@ def send_request(user, course_id, page, page_size, path="", text=None):
             params=params
         )
     except RequestException:
+        log.error("Failed to connect to edx-notes-api: url=%s, params=%s", url, str(params))
```
Can we use `.format`  instead %s ?

muzaffaryousaf commented on lms/djangoapps/edxnotes/helpers.py:
```diff
@@ -319,16 +325,23 @@ def get_notes(request, course, page=DEFAULT_PAGE, page_size=DEFAULT_PAGE_SIZE, t
     try:
         collection = json.loads(response.content)
     except ValueError:
+        log.error("Invalid response received from notes api: response_content=%s", response.content)
```
Can we specify the exact reason in logged error which can tell us the problem. i.e. Incorrect/invalid JSON or Failed to load JSON.

muzaffaryousaf commented on lms/djangoapps/edxnotes/helpers.py:
```diff
@@ -319,16 +325,23 @@ def get_notes(request, course, page=DEFAULT_PAGE, page_size=DEFAULT_PAGE_SIZE, t
+    expected_keys = ['total', 'rows', 'num_pages', 'start', 'next', 'previous', 'current_page']
     keys = collection.keys()
     if not keys or not all(key in expected_keys for key in keys):
         raise EdxNotesParseError(_("Invalid response received from notes api."))
```
Can we log this error too which can tell us the exact problem.

muzaffaryousaf commented on lms/static/js/edxnotes/collections/notes.js:
```diff
@@ -17,30 +27,26 @@ define([
-                        var chapter = note.get('chapter'),
-                            section = note.get('section'),
-                            unit = note.get('unit');
+                this.each(function (note) {
```
This method should not execute when we disable course structure tab. Can you please confirm this.

muhammad-ammar replied:
Confirmed. This will not be called when `Location in Course` tab is disabled.

muzaffaryousaf commented:
@muhammad-ammar I'm done with another pass and left some comments. Let me know when you are done with changes. Thanks.

explorerleslie commented:
:+1:

muhammad-ammar commented:
@muzaffaryousaf I have addressed your feedback. PTAL. Thanks!

chris-mike commented:
questions for @muhammad-ammar. 

1. Do you need to show pagination if the user has posted 10 or fewer notes?
2. When pagination shows (after 11 notes), it appropriately switches to " < 1 / 2 > ". I'd like to be able to click on the 2 as well as the >.  Is this possible?  B/c of the font contrast, I am more drawn to the number 2, than I am to the right arrow. Clicking on the number might mitigate future usability issues.

@muhammad-ammar Let me know when you address these questions and I'll review again.

@explorerleslie Let me know what you think of these issues. They *may* not be directly related to pagination, but I believe they are relevant to wayfinding.  I'll add as bugs, unless you tell me otherwise.

1. At the top, it says you can view notes by "recent activity." This looks like the only sortable option (assuming it's a sort)?  I see that search results show here. If it is the only one, why show any choices at all?  Alternatively, can we build in search terms here... since it seems to function as a filter mechanism, more so than a search.
2. When you search, you get a "search results" next to "recent activity."  When you "x" out search results (i.e., remove the filter), the search terms stay in the search box.  I would think that removing search results would also clear the words from the search box.
3. You can't edit notes or tags within "notes," rather you have to go back to courseware to edit even a tag).  Why not?

cptvitamin commented:
:+1:   Pagination works great.  There are lots of contrast issues in Student Notes, and ideally, we would have some notification to the user that the page content has updated.  Both of these are beyond the scope of this PR though.

muhammad-ammar commented:
@chris-mike I have made changes and updated the sandbox. Now paging footer will not be shown when we have single page of notes.
Regarding 2nd point, Given @cptvitamin comment, it would make more sense if UX team can look at edxnotes color contrast issues in a separate PR.

muhammad-ammar commented:
jenkins run bokchoy

muhammad-ammar commented:
jenkins run lettuce

muhammad-ammar commented:
jenkins run lettuce

chris-mike commented:
:+1:  Thanks for the changes. I'll move all the other usability notes to a different PR.

muhammad-ammar commented:
jenkins run lettuce

symbolist commented on common/test/acceptance/pages/common/paging.py:
```diff
@@ -63,3 +63,8 @@ def is_previous_page_button_enabled(self):
         return 'is-disabled' not in self.q(css=css).attrs('class')[0]
+
+    @property
+    def footer_visibility(self):
```
Since it returns a boolean indicating whether the footer is visible or not, it should be `footer_visible`.

symbolist commented on common/test/acceptance/tests/lms/test_lms_edxnotes.py:
```diff
@@ -398,6 +401,19 @@ def _add_default_notes(self, tags=None):
                 updated=datetime(2015, 1, 1, 1, 1, 1, 1).isoformat()
             ),
         ]
+        if extra_notes and isinstance(extra_notes, int):
```
Why wouldn't "if extra_notes > 0" work?

symbolist commented on common/test/acceptance/tests/lms/test_lms_edxnotes.py:
```diff
@@ -1005,6 +1063,290 @@ def test_open_note_when_accessed_from_notes_page(self):
+            total_pages=2
+        )
+
+        self.assert_viewed_event('Search Results')
```
Is this supposed to only trigger for the first page of results?

muhammad-ammar replied:
Currently it is only being triggered on first page. I am not sure about the requirements as edxnotes event document has no information regarding events behavior in pagination.

symbolist replied:
I have added a sub-task for this: https://openedx.atlassian.net/browse/TNL-4068. Please check with analytics team what the expected behaviour is for pagination.

symbolist commented on lms/static/js/edxnotes/views/notes_page.js:
```diff
@@ -34,20 +36,23 @@ define([
-            // Add the Tags model after the Course Structure model.
-            this.tabsCollection.push(tagsModel);
+                // Add the Tags model after the Course Structure model.
+                this.tabsCollection.push(tagsModel);
```
Pushing this into the collection should not depend on whehter `course_structure` tab is enabled or not. It should be a separate `if`.

symbolist commented:
@muhammad-ammar Are the "View notes by:" tabs supposed to show if the user has no notes?

symbolist commented on common/test/acceptance/tests/lms/test_lms_edxnotes.py:
```diff
@@ -850,6 +904,7 @@ def test_search_behaves_correctly(self):
         self.assert_viewed_event('Search Results')
         self.assert_search_event('note', 4)
 
+    @skip("scroll to tag functionality is removed")
```
I would prefer "disabled".

muhammad-ammar commented:
> Are the "View notes by:" tabs supposed to show if the user has no notes?

@symbolist nope

symbolist commented:
@muhammad-ammar I am seeing this on the sandbox.

muhammad-ammar commented:
@symbolist steps to reproduce?

muhammad-ammar commented:
<img width="1182" alt="screen shot 2016-01-25 at 3 18 42 pm" src="https://cloud.githubusercontent.com/assets/6767924/12548278/023df36e-c377-11e5-995c-ff8e6bdff579.png">

@symbolist If you are talking about tabs header shown then it is the default behavior, I have just confirmed with master.

symbolist commented on lms/djangoapps/edxnotes/helpers.py:
```diff
@@ -141,15 +137,13 @@ def preprocess_collection(user, course, collection):
     store = modulestore()
     filtered_collection = list()
     cache = {}
+    include_extra_info = settings.NOTES_DISABLED_TABS == []
```
This should be `include_path_info = 'course_structure' in settings.NOTES_DISABLED_TABS`.

muhammad-ammar commented:
jenkins run bokchoy

muhammad-ammar commented:
@symbolist I have addressed your feedback. Please continue the review.

symbolist commented on lms/djangoapps/edxnotes/helpers.py:
```diff
@@ -141,15 +137,13 @@ def preprocess_collection(user, course, collection):
     store = modulestore()
     filtered_collection = list()
     cache = {}
+    include_path_info = ('course_structure' not in settings.NOTES_DISABLED_TABS)
```
Should it not be the opposite of this?

muhammad-ammar replied:
if `course_structure ` is present in `settings.NOTES_DISABLED_TABS` then it means we don't need to calculate path info because `Location in Course` tab is disabled.

symbolist replied:
Ah, yes, which is why having a `NOTES_ENABLED_TABS` setting instead would have been simpler to read.

symbolist commented:
This looks good to go. :+1:

muzaffaryousaf commented:
:+1:  Once we get green build.
